### DISCUSSION TITLE
Has Pinocchio provided a function to compute the Jacobian of dynamics?

### DISCUSSION THREAD

**Discussion by zixinz990** [stack-of-tasks/pinocchio#2241] [Q&A]
I am a new user of Pinocchio. This is really an amazing library.

I used Pinocchio to calculate the whole-body dynamics $\dot x = f(x, u)$ for a humanoid robot. $x$ is $[q, \dot q]$, where $q$ is the state in generalized coordinates. Specifically, I did sth like:

```
// Compute the mass matrix
Matrix<T, Dynamic, Dynamic> M = pinocchio::crba(model, data, q).template cast<T>();
M.template triangularView<StrictlyLower>() = M.transpose().template triangularView<StrictlyLower>();

// Compute dynamic drift -- Coriolis, centrifugal, gravity
Matrix<T, Dynamic, 1> b = pinocchio::rnea(model, data, q, dq, Matrix<T, Dynamic, 1>::Zero(nv)).template cast<T>();

// Compute generalized acceleration
Matrix<T, Dynamic, 1> ddq = M.ldlt().solve(u - b); // [torso linear acc, torso angular acc, joint ddq], 6+n
```

Now, I am trying to get the Jacobian of the dynamics. In other words, I want $\partial f/\partial x$ and $\partial f/\partial u$. Does Pinocchio already provide some functions to do this?

**Comment by matheecs** [comment]
You may try to use CasADi & Pinocchio3 to get Jacobian of dynamics

**Reply by zixinz990** [reply]
Thank you for the reply. I just found several functions related to the derivative computation in Pinocchio, such as computeABADerivatives and computeRNEADerivatives. But I am not sure if they are also good for constrained dynamics. BTW, any specific reason for using CasADi?
